cirq/ops/projectors.py

```diff
@@ -0,0 +1,75 @@
+from typing import Any, Dict, Iterable, List, Tuple, TYPE_CHECKING, Union
+
+import numpy as np
+
+from cirq import value
+from cirq.ops import raw_types
+
+if TYPE_CHECKING:
+    import cirq
+
+
+@value.value_equality
+class Projector(raw_types.Gate):
+    """A non-unitary gate that projects onto one or more qubits."""
+
+    def __init__(self,
+                 projection_basis: Union[List[List[float]], np.ndarray],
+                 qid_shape: Tuple[int, ...] = (2,)):
+        """
+        Args:
+            projection_basis: a (2**num_qubits, p) matrix that lists the
+                projection vectors.
+            qid_shape: Specifies the dimension of the projection. The default is
+                2 (qubit).
+
+        Raises:
+            ValueError: If the basis vector is empty.
+        """
+        projection_array = np.asarray(projection_basis)
+
+        if len(projection_array.shape) != 2:
+            raise ValueError('The input projection_basis must be a 2D array')
+
+        if np.prod(qid_shape) != projection_array.shape[1]:
+            raise ValueError(
+                "Invalid shape " +
+                f"{np.array(projection_array).shape} for qid_shape {qid_shape}")
+        self._projection_basis = projection_array
+        self._qid_shape = qid_shape
+
+    def _projection_basis_(self) -> np.ndarray:
+        return self._projection_basis
+
+    def _qid_shape_(self) -> Tuple[int, ...]:
+        return self._qid_shape
+
+    def _has_unitary_(self) -> bool:
+        return False
+
+    def _is_parameterized_(self) -> bool:
+        return False
+
+    def _channel_(self) -> Iterable[np.ndarray]:
+        A = self._projection_basis
+        AH = np.transpose(np.conjugate(A))
+        result = np.matmul(np.matmul(AH, np.linalg.inv(np.matmul(A, AH))), A)
```

Comment:
I'm not quite sure I follow. I think in my notation:
`
A = V^\dagger
`

I used this convention since it's easier to feed inputs. So in short, my formula is:
`
V (V^\dagger V)^-1 V^\dagger
`

In case the basis is made of orthogonal vectors, it does indeed reduce to your formula:
`
V^\dagger V = I_p
`

But the formula that I have is more general, because I don't think your reasoning works with non-orthogonal vector.

If the basis has p=d vectors, then yes the formula reduces to the identity, but that's kind of expected, right? The simplification that you have does work, but in the general case A/V is not square and thus cannot be inverted, right?

I just pulled the formula from Wikipedia, by the way:
https://en.wikipedia.org/wiki/Projection_(linear_algebra)#Formulas

Reply:
You're right! Your algorithm is more general than the suggestion, so let's keep as is (I have a different comment though I'll add separately).